build.gradle

```diff
@@ -1,68 +1,98 @@
-buildscript {
-	dependencies {
-		classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.4.0'
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import org.jetbrains.kotlin.gradle.dsl.JvmTarget
+
+plugins {
+	alias(libs.plugins.kotlin.jvm)
+	alias(libs.plugins.spotless)
+}
+
+kotlin {
+	jvmToolchain(21)
+}
+
+tasks.withType(KotlinCompile).configureEach {
+	compilerOptions {
+		jvmTarget = JvmTarget.JVM_11
 	}
 }
 
-apply plugin: 'org.jetbrains.kotlin.jvm'
+tasks.withType(JavaCompile).configureEach {
+	options.release.set(11)
+}
+
+spotless {
+	kotlin {
+		ktlint().editorConfigOverride([
+			'ktlint_standard_filename': 'disabled',
+		])
+		target('**/*.kt')
+	}
+}
 
 configurations {
 	r8
 }
 
 dependencies {
-	testImplementation 'junit:junit:4.13.2'
-	testImplementation 'com.google.truth:truth:1.1.3'
+	testImplementation(platform(libs.junit.bom))
+	testImplementation libs.junit
+	testImplementation libs.truth
 
-	r8 'com.android.tools:r8:2.0.99'
+	testRuntimeOnly(libs.junit.platform.launcher)
+	testRuntimeOnly(libs.junit.vintage.engine)
+
+	r8 libs.r8
 }
 
 def fatJarProvider = tasks.register('fatJar', Jar) { task ->
 	task.dependsOn(configurations.named('runtimeClasspath'))
 	task.dependsOn(tasks.named('jar'))
 
-	task.classifier 'fat'
+	task.archiveClassifier.set('fat')
 
 	task.manifest {
 		attributes 'Main-Class': 'com.jakewharton.gradle.dependencies.DependencyTreeDiff'
 	}
 
 	def sourceClasses = sourceSets.main.output.classesDirs
 	task.inputs.files(sourceClasses)
-
-	task.doFirst {
-		task.from files(sourceClasses)
-		task.from configurations.runtimeClasspath.asFileTree.files.collect { zipTree(it) }
-		task.exclude '**/*.kotlin_metadata'
-		task.exclude '**/*.kotlin_module'
-		task.exclude '**/*.kotlin_builtins'
-		task.exclude '**/module-info.class'
-		task.exclude 'META-INF/maven/**'
+	task.from files(sourceClasses)
+	task.from configurations.named('runtimeClasspath').map {
+		it.asFileTree.files.collect { it.isDirectory() ? it : zipTree(it) }
 	}
+
+	task.exclude '**/*.kotlin_metadata'
+	task.exclude '**/*.kotlin_module'
+	task.exclude '**/*.kotlin_builtins'
+	task.exclude '**/module-info.class'
+	task.exclude 'META-INF/maven/**'
 }
 
-def r8File = new File("$buildDir/libs/$archivesBaseName-r8.jar")
+def r8File = layout.buildDirectory.file("libs/${base.archivesName.get()}-r8.jar").get().asFile
+def rulesFile = project.file("src/main/rules.pro")
+
 def r8Jar = tasks.register('r8Jar', JavaExec) { task ->
-	def rules = file('src/main/rules.txt')
-	task.dependsOn(configurations.named('runtimeClasspath'))
-	task.inputs.files(fatJarProvider.get().outputs.files, rules)
+	task.javaLauncher.set(javaToolchains.launcherFor { spec ->
+		spec.languageVersion.set(JavaLanguageVersion.of(17))
+	})
```

Comment:
R8 isn't compatible with Java 21 yet, I used previous LTS version since it is already installed in CI pipeline and used by test tasks.